File: app/src/main/java/com/appchamp/wordchunks/ui/main/MainFragment.java
package com.appchamp.wordchunks.ui.main;


import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.appchamp.wordchunks.R;
import com.appchamp.wordchunks.data.LevelsRealmHelper;

import io.realm.Realm;

import static com.appchamp.wordchunks.util.Constants.LEVEL_STATE_CURRENT;


public class MainFragment extends Fragment {

    //private SmallBang smallBang;

    private OnMainFragmentClickListener callback;

    static MainFragment newInstance() {
        return new MainFragment();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.frag_main, container, false);

        // smallBang = SmallBang.attach2Window(getActivity());

        //ImageView kittyView = (ImageView) rootView.findViewById(R.id.imgKitty);
        //kittyView.setOnClickListener(v -> smallBang.bang(v));
        // kittyView.callOnClick();

        rootView.findViewById(R.id.imgSettingsIcon).setOnClickListener(
                (View v) -> callback.showSlidingMenu());
        rootView.findViewById(R.id.imgShareIcon).setOnClickListener(this::onShareClick);

        rootView.findViewById(R.id.btnPlay).setOnClickListener(this::onPlayClick);
        rootView.findViewById(R.id.btnDaily).setOnClickListener(this::onDailyClick);
        rootView.findViewById(R.id.btnPacks).setOnClickListener(this::onPacksClick);
        rootView.findViewById(R.id.btnStore).setOnClickListener(this::onStoreClick);
        return rootView;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        // This makes sure that the container activity has implemented
        // the callback interface. If not, it throws an exception
        try {
            callback = (OnMainFragmentClickListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString()
                    + " must implement OnMainFragmentClickListener");
        }
    }

    private void onPlayClick(View v) {
        Realm realm = Realm.getDefaultInstance();
        String levelId = getLastCurrentLevelId(realm);
        realm.close();
        if (levelId != null) {
            callback.startGameActivity(levelId);
        } else {
            // If all levels and packs were solved, showing the fragment
            callback.showGameFinishedFragment();
        }
    }

    private void onPacksClick(View v) {
        callback.showPacksActivity();
    }

    private void onStoreClick(View v) {
    }

    private void onDailyClick(View v) {
    }

    private void onShareClick(View v) {
    }

    private String getLastCurrentLevelId(Realm realm) {

        long countLevelsByCurrentState =
                LevelsRealmHelper.countLevelsByState(realm, LEVEL_STATE_CURRENT);

        // If not all levels were solved
        if (countLevelsByCurrentState != 0) {
            // Getting the last "current" level id
            return LevelsRealmHelper.findLastLevelByState(realm, LEVEL_STATE_CURRENT).getId();
        } else {
            // If all levels and packs were solved, showing the fragment
            return null;
        }
    }

}

File: app/src/main/java/com/appchamp/wordchunks/ui/main/MainFragment.kt
/*
 * Copyright 2017 Julia Kozhukhovskaya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.appchamp.wordchunks.ui.main

import android.arch.lifecycle.LifecycleFragment
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.appchamp.wordchunks.R
import com.appchamp.wordchunks.ui.customviews.StoreDialog
import com.appchamp.wordchunks.ui.game.GameActivity
import com.appchamp.wordchunks.ui.packs.PacksActivity
import com.appchamp.wordchunks.util.Constants
import kotlinx.android.synthetic.main.frag_main.*
import org.jetbrains.anko.clearTop
import org.jetbrains.anko.intentFor
import org.jetbrains.anko.share
import xyz.hanks.library.SmallBang


class MainFragment : LifecycleFragment() {
    private val TAG: String = javaClass.simpleName
    private val smallBang by lazy { SmallBang.attach2Window(activity) }
    private val viewModel by lazy { ViewModelProviders.of(activity).get(MainViewModel::class.java) }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater?.inflate(R.layout.frag_main, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        btnShare.setOnClickListener { onShareClick() }
        btnPlay.setOnClickListener { onPlayClick() }
        btnDaily.setOnClickListener { onDailyClick() }
        btnPacks.setOnClickListener { startPacksActivity() }
        btnStore.setOnClickListener { onStoreClick() }
        blur.setOnClickListener { smallBang.bang(it) }
    }

    private fun onShareClick() {
        activity.share(getString(R.string.share_text), getString(R.string.share_subject))
    }

    private fun onPlayClick() {
        val levelId = viewModel.getLevel()?.id
        if (levelId != null) {
            startGameActivity(levelId)
        } else {
            // All levels and packs were solved, show dialog
            Toast.makeText(
                    context,
                    "Congratulations! You have solved all of the levels!",
                    Toast.LENGTH_LONG)
                    .show()
        }
    }

    private fun onStoreClick() {
        val dialog = StoreDialog.newInstance()
        dialog.show(activity.supportFragmentManager, "fragment_store")
    }

    private fun onDailyClick() {
        tvDailyBadge.visibility = View.INVISIBLE
        viewModel.fetchDailyLevel()
        viewModel.isRealmLoaded().observe(this, Observer {
            if (it!!) {
                viewModel.getDailyPuzzleLevelId()?.let { startGameActivity(it) }
            }
        })
    }

    private fun startGameActivity(levelId: String) {
        startActivity(activity.intentFor<GameActivity>(
                Constants.EXTRA_LEVEL_ID to levelId).clearTop())
        activity.overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left)
    }

    private fun startPacksActivity() {
        startActivity(activity.intentFor<PacksActivity>().clearTop())
        activity.overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left)
    }
}